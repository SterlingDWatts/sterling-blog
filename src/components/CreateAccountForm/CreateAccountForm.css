.CreateAccountForm {
  padding: 8px 16px;
  max-width: 944px;
  margin: 0 auto;
}

.CreateAccountForm h2 {
  margin: 32px 0px 0px;
  font-size: 32px;
  font-weight: 600;
  text-align: center;
}

.CreateAccountForm .hint {
  margin: 0px 0px 32px;
  text-align: center;
}

.CreateAccountForm__buttons {
  max-width: 288px;
  margin: 0px auto;
}
