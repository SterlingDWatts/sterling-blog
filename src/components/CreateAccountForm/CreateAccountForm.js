import React, { Component } from "react";
import {
  Input,
  Button,
  Label,
  LabelGroup,
  ValidationError,
  Required
} from "../Utils/Utils";
import "./CreateAccountForm.css";

class CreateAccountForm extends Component {
  state = {
    firstName: {
      value: "",
      touched: false
    },
    lastName: {
      value: "",
      touched: false
    },
    username: {
      value: "",
      touched: false
    },
    email: {
      value: "",
      touched: false
    },
    confirmEmail: {
      value: "",
      touched: false
    },
    password: {
      value: "",
      touched: false
    },
    confirmPassword: {
      value: "",
      touched: false
    }
  };

  handleFirstNameChange = firstName => {
    this.setState({
      firstName: { value: firstName, touched: true }
    });
  };

  validateFirstName = () => {
    const firstName = this.state.firstName.value;
    const NAME_REGEX = /[A-z-]+/;
    if (firstName.length === 0) {
      return "First Name is required";
    } else if (firstName.startsWith(" ") || firstName.endsWith(" ")) {
      return "First Name cannot start or end with an empty space";
    } else if (!NAME_REGEX.test(firstName)) {
      return "First Name must only include letters and -";
    }
  };

  handleLastNameChange = lastName => {
    this.setState({
      lastName: { value: lastName, touched: true }
    });
  };

  validateLastName = () => {
    const lastName = this.state.lastName.value;
    const NAME_REGEX = /[A-z-]+/;
    if (lastName.length === 0) {
      return "Last Name is required";
    } else if (lastName.startsWith(" ") || lastName.endsWith(" ")) {
      return "Last Name cannot start or end with an empty space";
    } else if (!NAME_REGEX.test(lastName)) {
      return "Last Name must only include letters and -";
    }
  };

  handleUsernameChange = username => {
    this.setState({
      username: { value: username, touched: true }
    });
  };

  validateUsername = () => {
    const username = this.state.username.value;
    const USERNAME_REGEX = /([A-Za-z0-9.\-_]+)/;
    if (username.length === 0) {
      return "Username is required";
    } else if (username.length < 6 || username.length > 32) {
      return "Username must be between 6 and 32 characters";
    } else if (username.startsWith(" ") || username.endsWith(" ")) {
      return "Username cannot start or end with an empty space";
    } else if (!USERNAME_REGEX.test(username)) {
      return "Username must only include letters, numbers, ., -, and _";
    }
  };

  handleEmailChange = email => {
    this.setState({
      email: { value: email, touched: true }
    });
  };

  validateEmail = () => {
    const email = this.state.email.value;
    const EMAIL_REGEX = /([a-zA-Z0-9_\-.]+)@([a-zA-Z0-9_\-.]+)\.([a-zA-Z]{2,5})/;
    if (email.length === 0) {
      return "Email is required";
    } else if (!EMAIL_REGEX.test(email)) {
      return "Email must be valid";
    }
  };

  handleConfirmEmailChange = confirmEmail => {
    this.setState({
      confirmEmail: { value: confirmEmail, touched: true }
    });
  };

  validateConfirmEmail = () => {
    const confirmEmail = this.state.confirmEmail.value;
    if (confirmEmail.length === 0) {
      return "Please confirm Email";
    } else if (confirmEmail !== this.state.email.value) {
      return "Emails must match";
    }
  };

  handlePasswordChange = password => {
    this.setState({
      password: { value: password, touched: true }
    });
  };

  validatePassword = () => {
    const password = this.state.password.value;
    const PASSWORD_REGEX = /(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[!@#$%^&])[\S]+/;
    if (password.length === 0) {
      return "Password is required";
    } else if (!PASSWORD_REGEX.test(password)) {
      return "Password must include at least upper case, lower case, number, and special character";
    } else if (password.startsWith(" ") || password.endsWith(" ")) {
      return "Password must not start or end with an empty space";
    }
  };

  handleConfirmPasswordChange = confirmPassword => {
    this.setState({
      confirmPassword: { value: confirmPassword, touched: true }
    });
  };

  validateConfirmPassword = () => {
    const confirmPassword = this.state.confirmPassword.value;
    if (confirmPassword.length === 0) {
      return "Please Validate Password";
    } else if (confirmPassword !== this.state.password.value) {
      return "Passwords must match";
    }
  };

  render() {
    const firstNameError = this.validateFirstName();
    const lastNameError = this.validateLastName();
    const usernameError = this.validateUsername();
    const emailError = this.validateEmail();
    const confirmEmailError = this.validateConfirmEmail();
    const passwordError = this.validatePassword();
    const confirmPasswordError = this.validateConfirmPassword();
    return (
      <form className="CreateAccountForm">
        <h2>Create Account</h2>
        <div className="hint">
          <Required /> required fields
        </div>
        <LabelGroup className="CreateAccountForm__first_name">
          <Label htmlFor="first_name">
            First Name <Required />
          </Label>
          <Input
            name="first_name"
            id="first_name"
            required
            type="text"
            value={this.state.firstName.value}
            onChange={e => this.handleFirstNameChange(e.target.value)}
          />
          {this.state.firstName.touched && (
            <ValidationError message={firstNameError} />
          )}
        </LabelGroup>
        <LabelGroup className="CreateAccountForm__last_name">
          <Label htmlFor="last_name">
            Last Name <Required />
          </Label>
          <Input
            name="last_name"
            id="last_name"
            required
            type="text"
            value={this.state.lastName.value}
            onChange={e => this.handleLastNameChange(e.target.value)}
          />
          {this.state.lastName.touched && (
            <ValidationError message={lastNameError} />
          )}
        </LabelGroup>
        <LabelGroup className="CreateAccountForm__username">
          <Label htmlFor="username">
            Username <Required />
          </Label>
          <Input
            name="username"
            id="username"
            required
            type="text"
            value={this.state.username.value}
            onChange={e => this.handleUsernameChange(e.target.value)}
          />
          {this.state.username.touched && (
            <ValidationError message={usernameError} />
          )}
        </LabelGroup>
        <LabelGroup className="CreateAccountForm__email">
          <Label htmlFor="email">
            Email <Required />
          </Label>
          <Input
            name="email"
            id="email"
            required
            type="email"
            value={this.state.email.value}
            onChange={e => this.handleEmailChange(e.target.value)}
          />
          {this.state.email.touched && <ValidationError message={emailError} />}
        </LabelGroup>
        <LabelGroup className="CreateAccountForm__confirm_email">
          <Label htmlFor="confirm_email">
            Confirm Email <Required />
          </Label>
          <Input
            name="confirm_email"
            id="confirm_email"
            required
            type="email"
            value={this.state.confirmEmail.value}
            onChange={e => this.handleConfirmEmailChange(e.target.value)}
          />
          {this.state.confirmEmail.touched && (
            <ValidationError message={confirmEmailError} />
          )}
        </LabelGroup>
        <LabelGroup className="CreateAccountForm__password">
          <Label htmlFor="password">
            Password <Required />
          </Label>
          <Input
            name="password"
            id="password"
            required
            type="password"
            value={this.state.password.value}
            onChange={e => this.handlePasswordChange(e.target.value)}
          />
          {this.state.password.touched && (
            <ValidationError message={passwordError} />
          )}
        </LabelGroup>
        <LabelGroup className="CreateAccountForm__confirm_password">
          <Label htmlFor="confirm_password">
            Confirm Password <Required />
          </Label>
          <Input
            name="confirm_password"
            id="confirm_password"
            required
            type="password"
            value={this.state.confirmPassword.value}
            onChange={e => this.handleConfirmPasswordChange(e.target.value)}
          />
          {this.state.confirmPassword.touched && (
            <ValidationError message={confirmPasswordError} />
          )}
        </LabelGroup>
        <div className="CreateAccountForm__buttons">
          <Button
            type="submit"
            disabled={
              this.validateFirstName() ||
              this.validateLastName() ||
              this.validateUsername() ||
              this.validateEmail() ||
              this.validateConfirmEmail() ||
              this.validatePassword() ||
              this.validateConfirmPassword()
            }
          >
            Create Account
          </Button>
        </div>
      </form>
    );
  }
}

export default CreateAccountForm;
