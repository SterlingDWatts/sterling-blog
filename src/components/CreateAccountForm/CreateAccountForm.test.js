import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom"
import CreateAccountForm from "./CreateAccountForm";

describe("CreateAccountForm Component", () => {
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(
      <BrowserRouter>
        <CreateAccountForm />
      </BrowserRouter>,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  it("renders the UI as expected", () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <CreateAccountForm />
        </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
