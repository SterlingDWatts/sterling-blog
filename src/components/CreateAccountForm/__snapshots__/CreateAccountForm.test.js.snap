// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateAccountForm Component renders the UI as expected 1`] = `
<form
  className="CreateAccountForm"
>
  <h2>
    Create Account
  </h2>
  <div
    className="hint"
  >
    <span
      className="Required "
    >
      *
    </span>
     required fields
  </div>
  <div
    className="LabelGroup CreateAccountForm__first_name"
  >
    <label
      className="Label "
      htmlFor="first_name"
    >
      First Name 
      <span
        className="Required "
      >
        *
      </span>
    </label>
    <input
      className="Input "
      id="first_name"
      name="first_name"
      onChange={[Function]}
      required={true}
      type="text"
      value=""
    />
  </div>
  <div
    className="LabelGroup CreateAccountForm__last_name"
  >
    <label
      className="Label "
      htmlFor="last_name"
    >
      Last Name 
      <span
        className="Required "
      >
        *
      </span>
    </label>
    <input
      className="Input "
      id="last_name"
      name="last_name"
      onChange={[Function]}
      required={true}
      type="text"
      value=""
    />
  </div>
  <div
    className="LabelGroup CreateAccountForm__username"
  >
    <label
      className="Label "
      htmlFor="username"
    >
      Username 
      <span
        className="Required "
      >
        *
      </span>
    </label>
    <input
      className="Input "
      id="username"
      name="username"
      onChange={[Function]}
      required={true}
      type="text"
      value=""
    />
  </div>
  <div
    className="LabelGroup CreateAccountForm__email"
  >
    <label
      className="Label "
      htmlFor="email"
    >
      Email 
      <span
        className="Required "
      >
        *
      </span>
    </label>
    <input
      className="Input "
      id="email"
      name="email"
      onChange={[Function]}
      required={true}
      type="email"
      value=""
    />
  </div>
  <div
    className="LabelGroup CreateAccountForm__confirm_email"
  >
    <label
      className="Label "
      htmlFor="confirm_email"
    >
      Confirm Email 
      <span
        className="Required "
      >
        *
      </span>
    </label>
    <input
      className="Input "
      id="confirm_email"
      name="confirm_email"
      onChange={[Function]}
      required={true}
      type="email"
      value=""
    />
  </div>
  <div
    className="LabelGroup CreateAccountForm__password"
  >
    <label
      className="Label "
      htmlFor="password"
    >
      Password 
      <span
        className="Required "
      >
        *
      </span>
    </label>
    <input
      className="Input "
      id="password"
      name="password"
      onChange={[Function]}
      required={true}
      type="password"
      value=""
    />
  </div>
  <div
    className="LabelGroup CreateAccountForm__confirm_password"
  >
    <label
      className="Label "
      htmlFor="confirm_password"
    >
      Confirm Password 
      <span
        className="Required "
      >
        *
      </span>
    </label>
    <input
      className="Input "
      id="confirm_password"
      name="confirm_password"
      onChange={[Function]}
      required={true}
      type="password"
      value=""
    />
  </div>
  <div
    className="CreateAccountForm__buttons"
  >
    <button
      className="Button "
      disabled="First Name is required"
      type="submit"
    >
      Create Account
    </button>
  </div>
</form>
`;
